### PR TITLE
Move program load to ebpfsvc

### LINKED ISSUE #75: ebpfsvc: interface with verifier/jitter/ioctlsurface and abstraction layer
@saxena-anurag

### PR DESCRIPTION
**Pull request by saxena-anurag**
This PR contains the following:

1. Move the code to JIT compile / load the program to service.
2. ebpfapi.dll now makes RPC call to load the program.
3. Updated "Getting Started" doc with details about ebpfsvc.
4. Added mock code for RPC calls, to ensure end_to_end.exe works.

Fixes #75

### CODE REVIEW

File: ebpfapi/rpc_client.cpp
Comment (Alan-Jowett) [inline ebpfapi/rpc_client.cpp:24]: Don't use static endpoints. Use a dynamic endpoint to avoid spoofing.

File: ebpfapi/rpc_client.cpp
Comment (Alan-Jowett) [inline ebpfapi/rpc_client.cpp]: Don't have commented out code.

File: ebpfsvc/rpc_api.cpp
Comment (Alan-Jowett) [inline ebpfsvc/rpc_api.cpp:27]: Is this still needed? I thought the change to thread local for state avoided this now?

File: ebpfapi/rpc_client.cpp
Comment (Alan-Jowett) [inline ebpfapi/rpc_client.cpp:46]: You need to use mutual authentication to make sure you are talking to the "real" EBPF service.

File: ebpfsvc/rpc_util.cpp
Comment (Alan-Jowett) [inline ebpfsvc/rpc_util.cpp:10]: Don't use hard-coded endpoints. Use dynamic endpoints to avoid name squatting attacks.

File: libs/api/ebpf_api.cpp
Comment (Alan-Jowett) [inline libs/api/ebpf_api.cpp:25]: Logging?

File: libs/api/ebpf_api.cpp
Comment (Alan-Jowett) [inline libs/api/ebpf_api.cpp:232]: This memory is leaked if this function throws an exception. Why use calloc and not a vector?

**Review by Alan-Jowett — COMMENTED** [review]
Some initial comments. More as I review this.

File: ebpfapi/rpc_client.cpp
Comment (saxena-anurag) [reply ebpfapi/rpc_client.cpp]: already removed in the next commit.

File: include/ebpf_execution_type.h
Comment (dthaler) [inline include/ebpf_execution_type.h:15]: Curious why we need an entire header file for a single typedef.  Is there no other file that this can be combined with?

File: libs/api/Verifier.cpp
Comment (dthaler) [inline libs/api/Verifier.cpp:55]:
```suggestion
        // Check if the raw program can fit in the supplied buffer.
```

File: libs/api/rpc_client.h
Comment (dthaler) [inline libs/api/rpc_client.h:18]: Shouldn't the return type be an enum like ebpf_result_t?  Or is that part of another github issue?  if the latter then can you add a // comment with a TODO that references the github issue #?

File: libs/api_common/api_common.cpp
Comment (dthaler) [inline libs/api_common/api_common.cpp:90]:
```suggestion
    uint32_t result = invoke_ioctl(device_handle, request, reply);
```
avoid abbreviations

File: libs/api_common/api_common.cpp
Comment (dthaler) [inline libs/api_common/api_common.cpp:76]: can this return ebpf_result_t?  (even if it means you have to call windows_error_to_ebpf_result before returning)

File: libs/api_common/device_helper.cpp
Comment (dthaler) [inline libs/api_common/device_helper.cpp:16]: remove commented out code

File: libs/api_common/device_helper.cpp
Comment (dthaler) [inline libs/api_common/device_helper.cpp:20]: should this return ebpf_result_t?

File: libs/api_common/map_descriptors.cpp
Comment (dthaler) [inline libs/api_common/map_descriptors.cpp:96]: I think emplace_back would be more efficient than push_back here.  Any time you call a constructor + push_back, it should instead call emplace_back.   The difference is that emplace_back avoids a copy after construction whereas push_back does not.

File: libs/service/api_service.cpp
Comment (dthaler) [inline libs/service/api_service.cpp:20]:
```suggestion
#define MAX_CODE_SIZE_IN_BYTES (32 * 1024) // 32 KB
```

File: libs/service/api_service.cpp
Comment (dthaler) [inline libs/service/api_service.cpp:52]:
```suggestion
    for (const auto& [helper_id, address]: helper_id_to_adddress) {
        request->helper_id[index++] = helper_id;
```
use structured bindings instead of .first and .second

File: libs/service/api_service.cpp
Comment (dthaler) [inline libs/service/api_service.cpp:62]:
```suggestion
    for (auto& [helper_id, address]: helper_id_to_adddress) {
        helper.address = reply->address[index++];
```

File: libs/service/api_service.cpp
Comment (dthaler) [inline libs/service/api_service.cpp:76]: move windows_error_to_ebpf_result call inside the invoke_ioctl() method

File: libs/service/api_service.cpp
Comment (dthaler) [inline libs/service/api_service.cpp:118]:
```suggestion
    if (map_handles.empty()) {
```

File: libs/service/api_service.cpp
Comment (dthaler) [inline libs/service/api_service.cpp:257]:
```suggestion
```

File: libs/service/api_service.cpp
Comment (dthaler) [inline libs/service/api_service.cpp:255]: make this use ebpf_result_t instead of int

File: libs/service/api_service.cpp
Comment (dthaler) [inline libs/service/api_service.cpp:321]: see earlier comments about .first and .second.  And look for anywhere else in this PR that uses them.

File: include/ebpf_execution_type.h
Comment (saxena-anurag) [reply include/ebpf_execution_type.h:15]: This was earlier included in ebpf_api.h. And I had to include ebpf_api.h in any file where I only wanted ebpf_execution_type_t definition. Hence I moved it to a different header file, to reduce unnecessary includes.

File: libs/service/api_service.cpp
Comment (saxena-anurag) [reply libs/service/api_service.cpp:76]: I think same error code from invoke_ioctl may need to be converted to a different ebpf_result, depending on the situation.

Example:
When invoke_ioctl fails in case of EBPF_OPERATION_LOAD_CODE, "EBPF_PROGRAM_LOAD_FAILED" is returned.
When invoke_ioctl fails in case of any map operation, the app will get some different ebpf_result error code (like KEY_NOT_FOUND)?

File: libs/api/rpc_client.h
Comment (saxena-anurag) [reply libs/api/rpc_client.h:18]: Currently the ebpf_api_* APIs return uint32_t instead of ebpf_result_t. I can fix this when I start changing the APIs.

File: ebpfapi/rpc_client.cpp
Comment (saxena-anurag) [reply ebpfapi/rpc_client.cpp]: Changed to dynamic endpoints.

File: ebpfsvc/rpc_api.cpp
Comment (saxena-anurag) [reply ebpfsvc/rpc_api.cpp]: This is not needed now. Removed.

File: ebpfsvc/rpc_util.cpp
Comment (saxena-anurag) [reply ebpfsvc/rpc_util.cpp]: Fixed.

File: libs/api/ebpf_api.cpp
Comment (saxena-anurag) [reply libs/api/ebpf_api.cpp]: changed to vector and added a try-catch block.

File: libs/api_common/map_descriptors.cpp
Comment (saxena-anurag) [reply libs/api_common/map_descriptors.cpp]: changed to emplace_back

File: libs/api_common/device_helper.cpp
Comment (shankarseal) [inline libs/api_common/device_helper.cpp:3]: A top level comment on what this module is doing is helpful.

File: libs/api_common/device_helper.cpp
Comment (saxena-anurag) [reply libs/api_common/device_helper.cpp:21]: initialize_device_handle() is called by ebpf_service_initialize() and ebpf_api_initiate() both of which return uint32_t.

File: libs/service/api_service.cpp
Comment (saxena-anurag) [reply libs/service/api_service.cpp:263]: This function returns ebpf_result_t. "int error" is used only to get return code from verifier functions which return 0 or 1.

File: libs/api/rpc_client.h
Comment (dthaler) [reply libs/api/rpc_client.h:18]: It would be good if we can stop adding more debt and at least make any new additions use ebpf_result_t instead of adding more code we know is "wrong" (I gave the same feedback on Shankar's PR btw).

File: libs/service/api_service.cpp
Comment (dthaler) [reply libs/service/api_service.cpp:76]: Ok that's an acceptable answer as far as invoke_ioctl returning a win32 error code.  But then line 76 would be wrong because it doesn't take the request type into account.

File: libs/service/api_service.cpp
Comment (dthaler) [reply libs/service/api_service.cpp:263]: I'm saying returning 0 or 1 from analyze() in verifier_service.cpp is undesirable, and it should instead return ebpf_result_t from that function.

File: libs/api/rpc_client.h
Comment (saxena-anurag) [reply libs/api/rpc_client.h:18]: Agree. One question regarding guidance: Is it OK for internal functions use uint32_t and only the public APIs (and RPC APIs) return ebpf_result_t? 

Example, in this function (ebpf_rpc_verify_program()), this function returns RPC exception code if there is an RPC exception. If we change this function to return ebpf_result_t, rpc exception code will get absorbed.

Comment (dthaler) [reply libs/api/rpc_client.h:18]: uint32_t is a really bad type to use for things with discrete values that could be an enum, since they make debugging and maintainability far more difficult.  So one should always use an enum type instead of uint32_t even for internal functions, whenever possible.

File: libs/api/ebpf_api.cpp
Comment (saxena-anurag) [reply libs/api/ebpf_api.cpp]: Yeah logging is missing from all the UM components. Filed #247 to add ETW tracing in all the UM components.

File: libs/service/api_service.cpp
Comment (saxena-anurag) [reply libs/service/api_service.cpp:76]: In this specific case (_resolve_ec_function), it is ok to directly convert the win32 error code to ebpf result (and not do any special conversion).

File: libs/service/api_service.cpp
Comment (saxena-anurag) [reply libs/service/api_service.cpp:258]: Fixed the functions (_analyze() and few others). "error" in this function is now only used to get the win32 error from invoke_ioctl.

File: libs/api_common/device_helper.cpp
Comment (saxena-anurag) [reply libs/api_common/device_helper.cpp:23]: Also, as with invoke_ioctl(), initialize_device_handle() returns error code from CreateFile(), so I think it is better to directly return that error code from initialize_device_handle()?

File: libs/service/api_service.cpp
Comment (dthaler) [reply libs/service/api_service.cpp:76]: If you think so, yes.

File: libs/api_common/device_helper.cpp
Comment (dthaler) [reply libs/api_common/device_helper.cpp:23]: Is initialize_device_handle called from multiple places that would need different conversion?  If not, then use ebpf_result_t... push it down as far as possible.

File: libs/api/rpc_client.h
Comment (saxena-anurag) [reply libs/api/rpc_client.h]: changed these functions to return ebpf_result_t

File: libs/api_common/device_helper.cpp
Comment (saxena-anurag) [reply libs/api_common/device_helper.cpp]: Fixed.

File: include/ebpf_execution_type.h
Comment (dthaler) [reply include/ebpf_execution_type.h:15]: Where else is it included from?   This is ok for this PR but going forward we should combine it into some more general header rather than exploding the number of include files, I think.

File: docs/GettingStarted.md
Comment (dthaler) [inline docs/GettingStarted.md:35]:
```suggestion
* ebpfsvc.exe: A user-mode service that verifies and loads an eBPF program in the execution context.
```
nit: synonymous in British english, but subtle difference in American english.  Chicago Manual of Style says "that" is correct here (in American English)

File: ebpfapi/rpc_client.cpp
Comment (dthaler) [inline ebpfapi/rpc_client.cpp:62]:
```suggestion
    RPC_STATUS status =
```
can be combined like you did on line 80

File: include/ebpf_api.h
Comment (dthaler) [inline include/ebpf_api.h:60]: add `@retval` lines with common return values (but ok to do in a future PR if you prefer, since it was already bad before)

File: include/ebpf_api.h
Comment (dthaler) [inline include/ebpf_api.h:145]: same here (but ok to do in a future PR)

File: libs/service/api_service.cpp
Comment (dthaler) [inline libs/service/api_service.cpp:281]:
```suggestion
        // Verify the program.
```

File: libs/service/api_service.cpp
Comment (dthaler) [inline libs/service/api_service.cpp:393]:
```suggestion
    // This is needed to ensure the service can successfully start even
```

**Review by dthaler — APPROVED** [review]
Looks good modulo a couple of nits.  Recommend accepting suggestions and then I can merge.

File: include/ebpf_execution_type.h
Comment (saxena-anurag) [reply include/ebpf_execution_type.h:15]: This new header is also included in rpc_interface.idl. If I try to directly include ebpf_api.h (ebpf_execution_type_t was earlier defined in ebpf_api.h), then I get following compilation errors as ebpf_api.h defines some globals:

> Error	MIDL2001	instantiation of data is illegal; you must use "extern" or "static"

**Comment by saxena-anurag** [comment]
> Looks good modulo a couple of nits. Recommend accepting suggestions and then I can merge.

I will fix these, but I think Alan is still reviewing the RPC part. We can wait till Monday for him to review that part before merging.

File: include/ebpf_execution_type.h
Comment (dthaler) [reply include/ebpf_execution_type.h:15]: agree it should move out of ebpf_api.h.   PR #246 creates an include/ebpf_structs.h file which is probably a better name for this.

File: include/ebpf_api.h
Comment (saxena-anurag) [reply include/ebpf_api.h:60]: I can fix this in the next PR -- I am going to work on implementing the APIs next.

File: include/ebpf_api.h
Comment (saxena-anurag) [reply include/ebpf_api.h:145]: I can fix this in the next PR -- I am going to work on implementing the APIs next.

File: ebpfapi/rpc_client.cpp
Comment (saxena-anurag) [reply ebpfapi/rpc_client.cpp]: Filed #249 to track this.

File: ebpfapi/rpc_client.cpp
Comment (dthaler) [inline ebpfapi/rpc_client.cpp:15]: This push/pop doesn't wrap any include file, can lines 10-13 and 15 all be deleted or do they serve some purpose?

File: ebpfsvc/rpc_api.cpp
Comment (Alan-Jowett) [inline ebpfsvc/rpc_api.cpp:24]: Please annotate functions that won't throw exceptions with [noexcept](https://en.cppreference.com/w/cpp/language/noexcept_spec). This will make it clear when we assume no exceptions will be thrown. 

Does ebpf_verify_and_load_program throw exceptions? It's not clear.

File: ebpfapi/rpc_client.cpp
Comment (saxena-anurag) [reply ebpfapi/rpc_client.cpp]: Deleted these lines.

File: ebpfsvc/rpc_api.cpp
Comment (saxena-anurag) [reply ebpfsvc/rpc_api.cpp:24]: Fixed for some of the functions.

**Review by Alan-Jowett — APPROVED** [review]

**Pull request merged**
